### ISSUE TITLE
Support dynamic action_space A(s)

### ISSUE THREAD

**Issue by jeweinb** [DLR-RM/stable-baselines3#580]
I was wondering does this library support dynamically updating action_space during agent training? I need to put constraints on my model that simply disallows specific actions given a current state.

Right now I have code inside the step function does something like the below.

```
def step(a):
    next_state = model(current_state, a)
    self.action_space = action_dist(next_state)

```

I would expect that the agent would pick up the new action space and sample it on the next iteration. But it seems like baselines grabs ahold of action_space during init and stores it. Can you point me to the place in the code where baselines samples from the action_space we create inside init? I wonder if I can make some changes to the code to allow it to dynamically update action_space based on state.

**Comment by Miffyli** [comment]
This creation happens bit all around, but here is the part where the necessary network output is created for the distributions: https://github.com/DLR-RM/stable-baselines3/blob/master/stable_baselines3/common/policies.py#L523

You might want to take a look at invalid action masking, which you could use to create a big action space and then limit actions per step to limited amount. This was just merged in contrib repo: https://github.com/Stable-Baselines-Team/stable-baselines3-contrib/pull/25

**Comment by jeweinb** [comment]
Unfortunately it looks like that contrib only masks actions at the environment level. I need actions to be masked at a much more granular action (given state).

It also looks like there might be an even bigger issue at play here too. It looks like stablebaselines3 rebuilds the distribution from the n attribute of the MultiBinary gym space as a torch distribution. That isn't good because I override the MultiBinary gym space sample method to only return specific actions that I care about. So it looks like my assumption that stablebaselines3 uses the action_space.sample method is wrong... is that true?

**Comment by Miffyli** [comment]
> Unfortunately it looks like that contrib only masks actions at the environment level. I need actions to be masked at a much more granular action (given state).

Hmm it should work on per-state level (that is the idea: not all actions are available per step). I am not sure how the API works in that as I have not digged deeper into that implementation.

> So it looks like my assumption that stablebaselines3 uses the action_space.sample method is wrong... is that true?

Yeah stable-baselines3 does not use the `action_space.sample` method (it relies on its own distributions for random sampling and what have you).

**Comment by jeweinb** [comment]
Ok thank you for the quick validation. I think that is a pretty insidious switch that is happening in the code. I think there should be some warning up front in the documentation that you do not use the action_space.sample method. That is a really critical piece of information. Because I had an agent trained without error from my own custom MultiBinary space and I just happened to stumble on the code where you baselines is rebuilding the distribution from scratch.

**Comment by Miffyli** [comment]
Just so that I understood correctly:

You have a custom action space where `sample` does not sample actions uniformly. Instead it samples them, say, based on the state or biases towards some choices over others. You thought the SB3 code would use this function to randomly sample actions (i.e., include your weighting/mods), but it does not.

You can achieve such biasing by modifying the neural network architecture and the parameters within, but yeah, the environment's `sample` method will have no effect.

**Comment by jeweinb** [comment]
Yeah that is right. As I am looking at the action mask contrib package I think that might do what I need it to.

Other RL libraries make direct use of the sample method. So you might want to put a warning in the docs somewhere that make sure the users know baselines does not use it.

**Comment by araffin** [comment]
> Yeah stable-baselines3 does not use the action_space.sample method (it relies on its own distributions for random sampling and what have you).

SB3 uses the sample method for off-policy algorithm during "warmup phase", not for A2C/PPO or other on-policy algorithms.
I doubt this is different in other codebase.

**Comment by jeweinb** [comment]
You are right @araffin. I'm lefting wondering what is the purpose of even defining a distribution on the env in the first place, why not just require a dimension as input? Would it be possible as a design choice of baselines to allow users to directly define a baselines friendly distribution in the env? That would give me the flexibility to impose rules on the distributions output, while also allowing for on-policy algorithms to update the distribution and sample as appropriate.

**Comment by araffin** [comment]
> I'm lefting wondering what is the purpose of even defining a distribution on the env in the first place, why not just require a dimension as input? 

I think you are confusing action space and probability distribution.
But best answer is try implementing yourself an algorithm that work for any gym env if you have only the dimension.
Spoiler alert: this won't work for all cases (for instance, with `Discrete` actions, it is recommended to use a `Categorical` distribution even though the action will be an scalar at the end).

> Would it be possible as a design choice of baselines to allow users to directly define a baselines friendly distribution in the env? 

Probability distribution belongs to the algorithm in my opinion (which is different from action space). In fact, some algorithms (like TD3 and DQN in some extent) don't use any probability distribution.

> That would give me the flexibility to impose rules on the distributions output, while also allowing for on-policy algorithms to update the distribution and sample as appropriate.

you can fork SB3 and implement your own distribution, or use SB3 contrib and provide an action mask.

**Comment by jeweinb** [comment]
Ok, thanks for answering my questions. I'm pretty new to designing a real world RL algo and learning a lot. I have a solution using the SB3 contrib package.
